File: searchlib/src/main/java/com/yahoo/searchlib/aggregation/CountAggregationResult.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.searchlib.aggregation;

import com.yahoo.searchlib.expression.IntegerResultNode;
import com.yahoo.searchlib.expression.ResultNode;
import com.yahoo.vespa.objects.Deserializer;
import com.yahoo.vespa.objects.ObjectVisitor;
import com.yahoo.vespa.objects.Serializer;

/**
 * This is an aggregated result holding the number of aggregated hits.
 *
 * @author baldersheim
 * @author Simon Thoresen Hult
 */
public class CountAggregationResult extends AggregationResult {

    public static final int classId = registerClass(0x4000 + 81, CountAggregationResult.class);
    private long count = 0;

    /** Constructs an empty result node. <b>NOTE:</b> This instance is broken until non-optional member data is set. */
    public CountAggregationResult() {

    }

    /**
     * Constructs an instance of this class with given count value.
     *
     * @param count the initial number of hits
     */
    public CountAggregationResult(long count) {
        setCount(count);
    }

    /** Returns the number of aggregated hits. */
    public final long getCount() {
        return count;
    }

    /**
     * Sets the number of aggregated hits.
     *
     * @param count the count
     * @return this, to allow chaining
     */
    public final CountAggregationResult setCount(long count) {
        this.count = count;
        return this;
    }

    @Override
    public ResultNode getRank() {
        return new IntegerResultNode(count);
    }

    @Override
    protected void onMerge(AggregationResult result) {
        count += ((CountAggregationResult)result).count;
    }

    @Override
    protected int onGetClassId() {
        return classId;
    }

    @Override
    protected void onSerialize(Serializer buf) {
        super.onSerialize(buf);
        buf.putLong(null, count);
    }

    @Override
    protected void onDeserialize(Deserializer buf) {
        super.onDeserialize(buf);
        count = buf.getLong(null);
    }

    @Override
    protected boolean equalsAggregation(AggregationResult obj) {
        return count == ((CountAggregationResult)obj).count;
    }

    @Override
    public int hashCode() {
        return super.hashCode() + (int)count;
    }

    @Override
    public void visitMembers(ObjectVisitor visitor) {
        super.visitMembers(visitor);
        visitor.visit("count", count);
    }

}

File: searchlib/src/vespa/searchlib/aggregation/countaggregationresult.cpp
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "aggregation.h"

#include <vespa/document/fieldvalue/document.h>
#include <vespa/searchlib/expression/resultvector.h>

#include <vespa/searchlib/aggregation/aggregation.hpp>
#include <vespa/vespalib/objects/visit.hpp>

using search::expression::ResultNodeVector;
using vespalib::Deserializer;
using vespalib::Serializer;

namespace search::aggregation {

IMPLEMENT_AGGREGATIONRESULT(CountAggregationResult, AggregationResult);

CountAggregationResult::CountAggregationResult(uint64_t count) : AggregationResult(), _count(count) {
}

void CountAggregationResult::initForUnitTest(const ResultNode& result) {
    _count.set(result);
}

void CountAggregationResult::onPrepare(const ResultNode&) {
}

void CountAggregationResult::onReset() {
    setCount(0);
}

void SumAggregationResult::onReset() {
    _sum.reset(static_cast<NumericResultNode*>(_sum->getClass().create()));
}

void CountAggregationResult::onMerge(const AggregationResult& b) {
    _count.add(static_cast<const CountAggregationResult&>(b)._count);
}

void CountAggregationResult::onAggregate(const ResultNode& result) {
    if (result.isMultiValue()) {
        _count += static_cast<const ResultNodeVector&>(result).size();
    } else {
        ++_count;
    }
}

Serializer& CountAggregationResult::onSerialize(Serializer& os) const {
    AggregationResult::onSerialize(os);
    return _count.serialize(os);
}

Deserializer& CountAggregationResult::onDeserialize(Deserializer& is) {
    AggregationResult::onDeserialize(is);
    return _count.deserialize(is);
}

void CountAggregationResult::visitMembers(vespalib::ObjectVisitor& visitor) const {
    AggregationResult::visitMembers(visitor);
    visit(visitor, "count", _count);
}

} // namespace search::aggregation
